### PR TITLE
fix #43, #44, #46

### LINKED ISSUE #43: refactor updateTopicBalance etc func to async/await

### PR DESCRIPTION
**Pull request by frankobe**
separate this fix into two commits. 
1: refactor all func to async/await to better handle error
2: batch rpc request and ensure requests r fired batch after batch

### CODE REVIEW

File: src/sync/index.js [hunk 1]
```diff
@@ -3,7 +3,10 @@ const _ = require('lodash');
 const connectDB = require('../db')
 
 const Qweb3 = require('qweb3');
-const qclient = new Qweb3('http://bodhi:bodhi@localhost:13889');
+const Qweb3Contract = require('qweb3/src/contract');
+const QTUM_RPC_ADDRESS = 'http://bodhi:bodhi@localhost:13889';
+const qclient = new Qweb3(QTUM_RPC_ADDRESS);
+
```
Comment (chrisli30) [inline src/sync/index.js]: Put QTUM_RPC_ADDRESS in a config file instead of sync/index.js like bodhi-api did?

File: src/sync/index.js [hunk 2]
```diff
@@ -403,111 +408,126 @@ async function sync(db){
...
+      value = await contract.call('getTotalVotes', { methodArgs: [], senderAddress: senderAddress});
+    } catch(err){
+      console.error(`Error: getTotalVotes for oracle ${oracleAddress}, ${err.message}`);
+      resolve();
+      return;
+    }
+  }
+
+  let balances = _.map(value[0].slice(0, oracle.options.length), (balance_BN) => {
+    return balance_BN.toJSON();
```
Comment (chrisli30) [inline src/sync/index.js]: value and value[0] needs undefined check. Access undefined[0] or undefined.slice() will throw an error.

or even better _.isEmpty().

File: src/sync/index.js [hunk 3]
```diff
@@ -403,111 +408,126 @@ async function sync(db){
...
   });
+
+  try {
+    await db.Oracles.updateOne({address: oracleAddress}, { $set: { amounts: balances }});
+    resolve();
+  } catch(err){
```
Comment (chrisli30) [inline src/sync/index.js]: Is return needed after resolve()?

File: src/sync/index.js [hunk 4]
```diff
@@ -403,111 +408,126 @@ async function sync(db){
...
+  }catch(err){
+    console.error(`Error: getTotalBets for topic ${topicAddress}, ${err.message}`);
+    resolve();
+    return;
+  }
+
+  let totalBetsBalances = _.map(totalBetsValue[0].slice(0, topic.options.length), (balance_BN) =>{
+    return balance_BN.toJSON();
   });
+
```
Comment (chrisli30) [inline src/sync/index.js]: totalBetsValue and totalBetsValue[0] need undefined check
totalVotesValue is the same.

File: src/sync/index.js [hunk 5]
```diff
@@ -365,149 +369,164 @@ async function sync(db){
...
-          setTimeout(startSync, 5000);
+  },
+  async function(){
+    // let updateOraclesPassedEndBlockPromise = new Promise((resolve) => {
+    //   updateOraclesPassedEndBlock(currentBlockChainHeight, db, resolve);
+    // });
+    // var oraclesNeedBalanceUpdateArray = Array.from(oraclesNeedBalanceUpdate);
+    // let updateOraclePromiseStack = oraclesNeedBalanceUpdateArray.map(async (oracle_address) => {
+    //   await updateOracleBalance(oracle_address, topicsNeedBalanceUpdate, db);
+    // });
```
Comment (dwalintukan) [inline src/sync/index.js]: delete old code?

File: src/sync/index.js [hunk 6]
```diff
@@ -365,149 +369,164 @@ async function sync(db){
...
+
+    // updateOraclePromiseStack.push(updateOraclesPassedEndBlockPromise);
```
Comment (dwalintukan) [inline src/sync/index.js]: delete?

File: src/sync/index.js (hunk 1 above)
Comment (frankobe) [reply src/sync/index.js]: done

File: src/sync/index.js (hunk 2 above)
Comment (frankobe) [reply src/sync/index.js]: i think it is fine 4 now unless qweb3 (ethabi) turns unexpected result without error

File: src/sync/index.js (hunk 5 above)
Comment (frankobe) [reply src/sync/index.js]: done

File: src/sync/index.js (hunk 6 above)
Comment (frankobe) [reply src/sync/index.js]: done

**Pull request merged**